### ISSUE TITLE
Add version of Particle.findall which returns generator?

### ISSUE THREAD

**Issue by HDembinski** [scikit-hep/particle#312]
I love the API of `Particle.findall`. A minor nitpick for me is that it returns a list, while the modern way would be to return a generator.

I think it would be easy to change the implementation to return a generator. However, simply replacing the return type in this method with a generator would be a breaking change. So perhaps another method could be added that returns the generator, and `Particle.findall` internally just calls that. I am not sure how this should be called, but perhaps `Particle.finditer`?

**Comment by henryiii** [comment]
It is extremely important that the list retuned by `findall` is sorted. The "first" results are always the simplest; DecayLangauge uses this. However, we could make sure the original list of particles is sorted; then returning an iterator from `finditer` would be possible; as a bonus, this might be a bit faster too after the initial setup, since it doesn't have to sort after finding particles.

**Comment by henryiii** [comment]
(The other option would be for `findall` to return `sorted(finditer)`, but it would be much nicer to not have the ordering different between the two)

**Comment by eduardo-rodrigues** [comment]
> It is extremely important that the list retuned by `findall` is sorted. The "first" results are always the simplest; DecayLangauge uses this. However, we could make sure the original list of particles is sorted; then returning an iterator from `finditer` would be possible; as a bonus, this might be a bit faster too after the initial setup, since it doesn't have to sort after finding particles.

I like it too that the first results are the most relevant. As far as DecayLanguage is concerned the package no longer relies on ordering or findall, in fact. This is because DecayLanguage needs by construction to deal transparently with particles in the PDG *and many* other particles that are only known to EvtGen, newly defined particles, exotic beasts, etc. A while back I made the searches and "translations needed robust by using directly the bimaps I defined in Particle.

In short, we have no constraints from that side.

**Comment by eduardo-rodrigues** [comment]
Now, since we are discussing the API in non-backward compatible ways together with the suggestion to make `find` obsoleted, see https://github.com/scikit-hep/particle/issues/313 - if we go the way of providing only a method, eventually, then one could actually go the following way:
- Keep `findall` as-is for now.
- Break the API and have `find` return a generator for all search results.
- then remove findall since `find` would do as `findall` and then the distinction in name isn't needed anymore.

The one downside of this is that `find` would be totally different.

Note that the one different as of now between `findall` and `find` is that the latter throws an exception if more than one particle is found in the search. This we can get rid of if people are happy to check for the number of returned particles.

Let me know what you think. I dumped these thoughts fast so may have missed a point.

**Comment by henryiii** [comment]
> As far as DecayLanguage is concerned the package no longer relies on ordering or findall, in fact

The AmpGen portion still does. Plus it's really a useful thing to be able to take advantage of. The generator would not be ver useful if everyone had to completely iterate over it and manually sort.

> Then remove `findall`

Particle was designed to be easy to use on the command line. `list(Particle.find(...))` is much less friendly than `Particle.findall()` - you have to remember to start with `list` and have an extra set of parenthesis. Now, another possible solution would be to go the way Python 3 did with things like range, and make a generator object that also has a nicer repr, getitem, and len. But then you'd need to be careful with caching, or you'd end up running over the particles _more_ times instead of less. I'd rather keep `findall`, I think. But it could be done.

By the way, this name intentionally looks like `re.findall`, which is why `finditer` is attractive, there's also a `re.finditer`.

> we can get rid of if people are happy to check for the number of returned particles.

Changing the return type from a single particle to a list or generator is a breaking change. Let's not replace `Particle.find`, but maybe look toward the next major change.

If we rework this to break it into two classes, ParticleDB / Particle, then that's a good point to drop `.find` and leave only `.finditer/.findall`. Given the resemblance to `re`, I think we don't need to reintroduce `find`. However, if we do make this a smart generator, then we could just have one `find` that returns a smart generator. Again, I'd lean a bit toward simplicity.

**Comment by eduardo-rodrigues** [comment]
I will read both threads again since we're now discussing in 2 issues. What is then your proposal as a first step? I will try and get started on that asap.

**Comment by HDembinski** [comment]
I disagree with Henry. The only point of using the generator version is performance. If the generator version has to sort the results first, then do not implement it at all, because it would be a false promise.

My preference is to have a findall which returns whatever it returns now and a finditer that returns result without sorting, because sorting requires to build the full list of results anyway.

Sidenote: I don't understand why it has to be sorted. Let the user sort the results or make that optional. Im some situations it is nice to get a sorted list, but you should not force that performance hit upon every user. There are for sure use cases where the results do not need to be sorted.

**Comment by HDembinski** [comment]
Please consider the design principles I mentioned in the other thread, about functions which do the minimal amount of work, and about functions that, if they have to do extra work, then also return that.

In regard to finditer, that means if you have to make list internally anyway to sort some results and then by all means return the list and not a generator that yields items from the list.

**Comment by henryiii** [comment]
> If the generator version has to sort the results first, then do not implement it at all, because it would be a false promise.

I think we can pre-sort the internal list, and then always return the matches in order - that would eliminate the need for sorting after they are found.

> The only point of using the generator version is performance

If you have to exhaust the iterator to do anything useful, then you gain no performance. The list is small (and returns pre-existing objects), so there is no memory advantage and no significant performance difference between getting all matches and getting them one at a time if you have to get all of them anyway. The point of an iterator is to avoid having to exhaust it, or to chain together long computations to reduce memory - this is not the later, though, there are a limited number of entries to search.

>  I don't understand why it has to be sorted.

If the results are sorted, you can stop when you get particles that you are not interested in - often you don't care about exotic excited states. For example, if I do `particle.particle.Particle.findall("B")`, I get 322 entries. However, often, I'm really only interested in the first 1-20 or so matches - the rest of the more exotic particles, like `Be5`, I can ignore. If the iterator version can return this, that's a big savings for the iterator version.
